python/cucim/src/cucim/skimage/filters/_separable_filtering.py

```diff
@@ -0,0 +1,996 @@
+import math
+
+import cupy as cp
+import numpy as np
+
+from cucim.skimage._vendored import _ndimage_util as util
+from cucim.skimage._vendored._internal import _normalize_axis_index
+from cucim.skimage._vendored._ndimage_filters_core import (
+    _ndimage_CAST_FUNCTION, _ndimage_includes)
+
+if hasattr(math, 'prod'):
+    prod = math.prod
+else:
+    prod = np.prod
+
+
+def _get_constants(ndim, axis, kernel_size, anchor, patch_per_block=None):
+    if anchor is None:
+        anchor = kernel_size // 2
+    halo_pixels_needed = max(kernel_size - anchor, anchor)
+    if patch_per_block is None:
+        patch_per_block = 4
+
+    if ndim == 2:
+        # note, in 2d axis 0 = "y"
+        #             axis 1 = "x"
+        # for simplicity, keeping same halo size at both start and end
+        if axis == 1:
+            # as in OpenCV's column_filter.hpp
+            block_x = 16
+            block_y = 16
+            halo_size = math.ceil(halo_pixels_needed / block_x)
+        elif axis == 0:
+            # as in OpenCV's row_filter.hpp
+            block_x = 32  # 16 in CUDA example
+            block_y = 8  # 4 in CUDA example
+            halo_size = math.ceil(halo_pixels_needed / block_y)
+        # can have out of bounds access unless patch_per_block >= halo_size
+        patch_per_block = max(patch_per_block, halo_size)
+        block_z = 1
+    elif ndim == 3:
+        # note, in 3d axis 0 = "z"
+        #             axis 1 = "y"
+        #             axis 2 = "x"
+        # for simplicity, keeping same halo size at both start and end
+        if axis == 2:
+            # as in OpenCV's column_filter.hpp
+            block_x = 16
+            block_y = 4
+            block_z = 4
+            halo_size = math.ceil(halo_pixels_needed / block_x)
+        elif axis == 1:
+            # as in OpenCV's column_filter.hpp
+            block_x = 32
+            block_y = 4
+            block_z = 4
+            halo_size = math.ceil(halo_pixels_needed / block_y)
+        elif axis == 0:
+            # as in OpenCV's row_filter.hpp
+            block_x = 32
+            block_y = 4
+            block_z = 4
+            halo_size = math.ceil(halo_pixels_needed / block_z)
+        # can have out of bounds access unless patch_per_block >= halo_size
+        patch_per_block = max(patch_per_block, halo_size)
+    else:
+        raise NotImplementedError("TODO")
```

Comment:
Should there be more to this message?

Reply:
updated to mention only 2D and 3D are currently supported